File: pt5/1197.minimum-knight-moves.go
package main

func minKnightMoves(x int, y int) int {
	if x < 0 {
		x = -x
	}
	if y < 0 {
		y = -y
	}
	if x == 0 && y == 0 {
		return 0
	}

	type state struct {
		row      int
		col      int
		distance int
	}

	moves := [][2]int{{1, 2}, {2, 1}, {2, -1}, {1, -2}, {-1, -2}, {-2, -1}, {-2, 1}, {-1, 2}}
	queue := []state{{row: 0, col: 0, distance: 0}}
	visited := map[[2]int]bool{{0, 0}: true}

	for head := 0; head < len(queue); head++ {
		current := queue[head]
		for _, move := range moves {
			nextRow := current.row + move[0]
			nextCol := current.col + move[1]
			if nextRow == x && nextCol == y {
				return current.distance + 1
			}
			key := [2]int{nextRow, nextCol}
			if nextRow >= -2 && nextRow <= x+2 && nextCol >= -2 && nextCol <= y+2 && !visited[key] {
				visited[key] = true
				queue = append(queue, state{row: nextRow, col: nextCol, distance: current.distance + 1})
			}
		}
	}
	return -1
}

File: pt5/1197.minimum-knight-moves.java
import java.util.*;

class Solution {
    public int minKnightMoves(int x, int y) {
        x = Math.abs(x);
        y = Math.abs(y);
        if (x == 0 && y == 0) {
            return 0;
        }

        int[][] moves = {
            {1, 2}, {2, 1}, {2, -1}, {1, -2},
            {-1, -2}, {-2, -1}, {-2, 1}, {-1, 2}
        };
        Queue<int[]> queue = new ArrayDeque<>();
        Set<Long> visited = new HashSet<>();
        queue.offer(new int[] {0, 0, 0});
        visited.add(key(0, 0));

        while (!queue.isEmpty()) {
            int[] state = queue.poll();
            for (int[] move : moves) {
                int nr = state[0] + move[0];
                int nc = state[1] + move[1];
                if (nr == x && nc == y) {
                    return state[2] + 1;
                }
                long key = key(nr, nc);
                if (nr >= -2 && nr <= x + 2 && nc >= -2 && nc <= y + 2 && visited.add(key)) {
                    queue.offer(new int[] {nr, nc, state[2] + 1});
                }
            }
        }

        return -1;
    }

    private long key(int row, int col) {
        return (((long) row) << 32) ^ (col & 0xffffffffL);
    }
}

